msm/src/serialization/witness.rs

```diff
@@ -156,46 +160,119 @@ pub fn deserialize_field_element<Fp: Field, Env: InterpreterEnv<Fp>>(
-    env.bitmask_be(&input_limb1, 62, 47, c9);
-    env.bitmask_be(&input_limb1, 77, 62, c10);
-    {
+    let c6_var = {
+        let c6 = Env::get_column_for_msm_limb(6);
+        env.bitmask_be(&input_limb1, 17, 2, c6)
+    };
+    let c7_var = {
+        let c7 = Env::get_column_for_msm_limb(7);
+        env.bitmask_be(&input_limb1, 32, 17, c7)
+    };
+    let c8_var = {
+        let c8 = Env::get_column_for_msm_limb(8);
+        env.bitmask_be(&input_limb1, 47, 32, c8)
+    };
+    let c9_var = {
+        let c9 = Env::get_column_for_msm_limb(9);
+        env.bitmask_be(&input_limb1, 62, 47, c9)
+    };
+    let c10_var = {
+        let c10 = Env::get_column_for_msm_limb(10);
+        env.bitmask_be(&input_limb1, 77, 62, c10)
+    };
+    let c11_var = {
+        let c11 = Env::get_column_for_msm_limb(11);
         let res = (limbs[1] >> 77) & ((1 << (88 - 77)) - 1);
         let res_prime = limbs[2] & ((1 << 4) - 1);
         let res = res + (res_prime << (15 - 4));
         let res = Env::constant(res.into());
-        env.copy(&res, c11);
-    }
+        env.copy(&res, c11)
+    };
+
+    // Unfolding for readability.
+    // IMPROVEME using fold later.
+    // Shift left by 88
+    let shl_88 = Fp::from(1u128 << 88u128);
+    let mut shl_88_var = Env::constant(shl_88);
+
+    // b0 + b1 * 2^88
+    let mut constraint = limb0_var + limb1_var * shl_88_var.clone();
+    // Compute shift by 176
+    shl_88_var = shl_88_var * Env::constant(shl_88);
+
+    // b0 + b1 * 2^88 + b2 * 2^176
+    constraint = constraint + limb2_0_var * shl_88_var;
+
+    // Substract first 15 bit limb
+    // b0 + b1 * 2^88 + b2 * 2^176 - c0
+    constraint = constraint - c0_var;
+
+    // Substract the other decompositions in base 15
+    let mut cst = Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint.clone() - c1_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c2_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c3_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c4_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c5_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c6_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c7_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c8_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c9_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
+    constraint = constraint - c10_var * cst.clone();
+
+    cst = cst * Env::constant(Fp::from((1 << 15) as u64));
```

Comment:
you probably want to loop this

Reply:
Opening follow-up issue https://github.com/o1-labs/proof-systems/issues/1899. Will be done directly after merging other PRs to avoid conflicts if it is fine for you.